service/worker/scanner/data_corruption_workflow.go

```diff
@@ -0,0 +1,246 @@
+	"go.uber.org/cadence/workflow"
+	"go.uber.org/zap"
+
+	"github.com/uber/cadence/common/log/tag"
+
+	"github.com/uber/cadence/common/metrics"
+
+	c "github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/reconciliation"
+	"github.com/uber/cadence/common/reconciliation/entity"
+	"github.com/uber/cadence/common/reconciliation/fetcher"
+	"github.com/uber/cadence/common/reconciliation/invariant"
+	"github.com/uber/cadence/common/resource"
+)
+
+func init() {
+	workflow.RegisterWithOptions(
+		CheckDataCorruptionWorkflow,
+		workflow.RegisterOptions{Name: reconciliation.CheckDataCorruptionWorkflowType},
+	)
+	activity.Register(ExecutionFixerActivity)
+	activity.Register(EmitResultMetricsActivity)
+}
+
+const (
+	workflowTimer        = 5 * time.Minute
+	maxSignalNumber      = 1000
+	fixerActivityTimeout = time.Minute
+)
+
+// CheckDataCorruptionWorkflow is invoked by remote cluster via signals
+func CheckDataCorruptionWorkflow(ctx workflow.Context, fixList []entity.Execution) error {
+
+	logger := workflow.GetLogger(ctx)
+	signalCh := workflow.GetSignalChannel(ctx, reconciliation.CheckDataCorruptionWorkflowSignalName)
+	signalCount := 0
+
+	for {
+		selector := workflow.NewSelector(ctx)
+		// timer
+		timerCtx, timerCancel := workflow.WithCancel(ctx)
+		waitTimer := workflow.NewTimer(timerCtx, workflowTimer)
+		selector.AddFuture(waitTimer, func(f workflow.Future) {
+			// do nothing. Unblock the selector
+		})
+
+		// signal
+		selector.AddReceive(signalCh, func(c workflow.Channel, more bool) {
+			var fixExecution entity.Execution
+			for c.ReceiveAsync(&fixExecution) {
+				signalCount++
+				fixList = append(fixList, fixExecution)
+			}
+		})
+
+		selector.Select(ctx)
+		if len(fixList) == 0 {
+			return nil
+		}
+		timerCancel()
+
+		timeout := fixerActivityTimeout * time.Duration(len(fixList))
+		activityOptions = workflow.ActivityOptions{
+			ScheduleToStartTimeout: time.Minute,
+			StartToCloseTimeout:    timeout,
+			HeartbeatTimeout:       fixerActivityTimeout,
+			RetryPolicy:            &activityRetryPolicy,
+		}
+		activityCtx := workflow.WithActivityOptions(ctx, activityOptions)
+		var fixResults []invariant.FixResult
+		err := workflow.ExecuteActivity(activityCtx, ExecutionFixerActivity, fixList).Get(activityCtx, &fixResults)
+		if err != nil {
+			logger.Error("failed to run execution fixer activity", zap.Error(err))
+			return err
+		}
+
+		activityOptions = workflow.ActivityOptions{
+			ScheduleToStartTimeout: time.Minute,
+			StartToCloseTimeout:    time.Minute,
+		}
+		activityCtx = workflow.WithActivityOptions(ctx, activityOptions)
+		err = workflow.ExecuteActivity(activityCtx, EmitResultMetricsActivity, fixResults).Get(activityCtx, nil)
+		if err != nil {
+			logger.Error("failed to run execution fixer activity", zap.Error(err))
+			return err
+		}
+
+		fixList = []entity.Execution{}
+		workflow.GetMetricsScope(ctx)
+		if signalCount > maxSignalNumber {
```

Comment:
I guess this means changing the `maxSignalNumber` requires a GetVersion() call? Probably we should add a comment for this at where `maxSignalNumber` is defined?
Another idea is to use `workflow.SideEffect` to record the max number in history.